glommio/src/read_write_lock.rs

```diff
@@ -0,0 +1,1214 @@
+                        if *waiter_kind == WaiterKind::WRITER {
+                            break;
+                        }
+
+                        let (_, waker) = entry.remove();
+                        waker.wake();
+
+                        rw.waiters.pop_front().unwrap();
+                    }
+                    Vacant(_) => unreachable!(),
+                }
+            }
+        }
+    }
+    fn wake_up_readers_and_first_writer(rw: &mut State<T>) {
+        loop {
+            let waiter_id = rw.waiters.pop_front();
+            if let Some(waiter_id) = waiter_id {
+                let (waiter_kind, waker) = rw.waiters_map.remove(&waiter_id).unwrap();
+                waker.wake();
+
+                if waiter_kind == WaiterKind::WRITER {
+                    rw.queued_writers -= 1;
+                    break;
+                }
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn wake_up_all_fibers(rw: &mut State<T>) {
+        for (_, (_, waker)) in rw.waiters_map.drain() {
+            waker.wake();
+        }
+
+        rw.waiters.clear();
+    }
+}
+
+impl<T> Drop for ReadWriteLock<T> {
+    fn drop(&mut self) {
+        self.close();
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::read_write_lock::LockClosedError;
+    use crate::read_write_lock::ReadWriteLock;
+    use crate::read_write_lock::TryReadWriteLockError;
+
+    use crate::Semaphore;
+    use rand::{self, Rng};
+    use std::cell::RefCell;
+    use std::rc::Rc;
+
+    #[derive(Eq, PartialEq, Debug)]
+    struct NonCopy(i32);
+
+    #[test]
+    fn test_smoke() {
+        test_executor!(async move {
+            let lock = ReadWriteLock::new(());
+            drop(lock.read().await.unwrap());
+            drop(lock.write().await.unwrap());
+            drop((lock.read().await.unwrap(), lock.read().await.unwrap()));
+            drop(lock.read().await.unwrap());
+        });
+    }
+
+    #[test]
+    fn test_frob() {
+        test_executor!(async move {
+            const N: u64 = 10;
+            const M: usize = 1000;
+
+            let r = Rc::new(ReadWriteLock::new(()));
+            let mut futures = Vec::new();
+
+            for _ in 0..N {
+                let r = r.clone();
+
+                let f = Task::local(async move {
+                    let mut rng = rand::thread_rng();
+                    for _ in 0..M {
+                        if rng.gen_bool(1.0 / (N as f64)) {
+                            drop(r.write().await.unwrap());
+                        } else {
+                            drop(r.read().await.unwrap());
+                        }
+                    }
+                });
+
+                futures.push(f);
+            }
+
+            join_all(futures).await;
+        });
+    }
+
+    #[test]
+    fn test_close_wr() {
+        test_executor!(async move {
+            let rc = Rc::new(ReadWriteLock::new(1));
+            let rc2 = rc.clone();
+
+            Task::local(async move {
```

Comment:
If I read this correctly this test acquires the lock in write mode a task queue and tries to read it from another (the default one).

While this works, it happens to work due to an implementation detail that can change. Nothing guarantees which task queue runs first. And as a matter of fact I am surprised this works. I would expect the default tq to run first.

Where I need to test things like this I have been relying on explicit serialization: Essentially add a control variable inside a `Cell`, and a yielding loop `while status.get() == x { Local::later().await; }`

For an example of that you can look at the `permit_raii_works` in the semaphore tests.

Reply:
Hi, @glommer in this test at first we await that task will be finished first and once it is finished we check that error is returned as a result of the read. So as I understand `.await` at the end of the task creation guarantees this order, because of for me it looks like analogous of .join() for a normal thread. Am I mistaken?

Reply:
Ok. It's my bad. You are testing a different thing than what I thought so my brain mentally attached a `.detach()` there =)
(You knw wn you rd jst half a sentc but your brain fills the rest?)

You are testing that the read or write will fail because the lock was closed, not that the lock was held.

You are correct, the task will spawn but the `.await` statement is a synchronization point.

Reply:
> You knw wn you rd jst half a sentc but your brain fills the rest? 
I always skim the news articles and sometimes the next day I become surprised that article, when I read it again, says absolutely different thing :-D .

Reply:
Well in that case maybe they changed it as they declared war on Eastasia